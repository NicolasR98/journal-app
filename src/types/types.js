export const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error',
    uiStartLoading: '[UI] Start isLoading',
    uiStopLoading: '[UI] Stop isLoading',
}