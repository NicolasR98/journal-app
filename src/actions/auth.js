import { createUserWithEmailAndPassword, getAuth, signInWithPopup, updateProfile } from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebaseConfig';
import { types } from '../types/types'

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

// To handle an async function, the helper must return another function
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        setTimeout(() => {
            dispatch(login(123, 'nicolas'));
        }, 3500);
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        // Get auth and make a petition to firebase to create an user
        const auth = getAuth();
        createUserWithEmailAndPassword(auth, email, password)
            // Once created, make a post req to firebase to update the current
            // user displayName (async)
            .then(async ({ user }) => {
                await updateProfile(user, { displayName: name });
                dispatch(
                    login(user?.uid, user.displayName)
                );
            })
            .catch(err => err);
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
            .then(({ user }) => {
                dispatch(
                    login(user?.uid, user?.displayName)
                );
            })
            .catch(e => console.log(e));
    };
};