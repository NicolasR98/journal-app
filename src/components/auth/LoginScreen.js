import React from "react"
import { useDispatch, useSelector } from "react-redux";

import { Link } from "react-router-dom";
import isEmail from "validator/lib/isEmail";

import { startGoogleLogin, startLoginEmailPassword } from "../../actions/auth";
import { uiRemoveError, uiSetError } from "../../actions/ui";

import { useForm } from "../../hooks/useForm";

const initialState = {
    email: 'example@mail.com',
    password: '12345',
};

export const LoginScreen = () => {
    const dispatch = useDispatch();
    const { msgError, isLoading } = useSelector(state => state?.ui);

    const [formValues, handleInputChange] = useForm(initialState)
    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startLoginEmailPassword(email, password));
        }
    }

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    }

    const isFormValid = () => {
        if (email.length === 0 || password.length === 0) {
            dispatch(uiSetError('Fields cannot be empty'))
            return false;
        } else if (!isEmail(email)) {
            dispatch(uiSetError('Email is not valid'));
            return false;
        } else if (password.length < 5) {
            dispatch(uiSetError('Password should be at least 6 characters long'));
            return false;
        } else if (msgError) {
            dispatch(uiRemoveError());
        };

        return true;
    };

    return (
        <>
            <h3 className="auth__title">Login</h3>

            <form onSubmit={handleLogin}>
                {
                    msgError && (
                        <div className='auth__alert-error'>
                            {msgError}
                        </div>
                    )
                }
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    autoComplete="off"
                    value={password}
                    onChange={handleInputChange}
                />

                <button
                    className="btn btn-block btn-primary"
                    type="submit"
                    disabled={isLoading}
                >
                    Login
                </button>

                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div
                        className="google-btn"
                        onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img
                                className="google-icon"
                                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                                alt="google button"
                            />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link
                    className="link"
                    to="/auth/register"
                >
                    Create new account
                </Link>
            </form>
        </>
    );
};
