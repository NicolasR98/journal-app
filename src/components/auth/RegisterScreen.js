import React from 'react'
import { useDispatch, useSelector } from 'react-redux'

import { Link } from 'react-router-dom'

import isEmail from 'validator/lib/isEmail'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'

import { uiRemoveError, uiSetError } from '../../actions/ui'
import { useForm } from '../../hooks/useForm'

const initialState = {
    name: 'Hernando',
    email: 'example@mail.com',
    password: '12345',
    passwordConfirm: '12345',
};

export const RegisterScreen = () => {
    const dispatch = useDispatch();
    const { msgError, isLoading } = useSelector(state => state?.ui);

    const [formValues, handleInputChange] = useForm(initialState);
    const { name, email, password, passwordConfirm } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        };
    };

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(uiSetError('Name required'));
            return false;
        } else if (!isEmail(email)) {
            dispatch(uiSetError('Email is not valid'));
            return false;
        } else if (password !== passwordConfirm || password.length < 5) {
            dispatch(uiSetError('Password should be at least 6 characters long'));
            return false;
        } else {
            dispatch(uiRemoveError());
            return true;
        };
    };

    return (
        <>
            <h3 className="auth__title">Register</h3>

            <form 
                onSubmit={handleRegister}
                className='animate__animated animate__fadeIn animate__faster'
            >
                {
                    msgError && (
                        <div className='auth__alert-error'>
                            {msgError}
                        </div>

                    )
                }
                <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={name}
                />

                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={email}
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={password}
                />

                <input
                    type="password"
                    placeholder="Confirm password"
                    name="passwordConfirm"
                    className="auth__input"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={passwordConfirm}
                />

                <button
                    className="btn btn-block btn-primary mb-5"
                    type="submit"
                    disabled={isLoading}
                >
                    Register
                </button>

                <Link
                    className="link"
                    to="/auth/login"
                >
                    Already registered?
                </Link>
            </form>
        </>
    )
}
