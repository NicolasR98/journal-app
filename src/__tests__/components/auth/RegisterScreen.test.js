import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const msgError = 'Email is not valid';
const initialState = {
    auth: {},
    ui: {
        isLoading: false,
        msgError: null,
    },
};

let store = mockStore(initialState);

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
);

describe('Tests on <RegisterScreen />', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should display correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should fire the correct actions if empty email given', () => {
        const emailInput = wrapper.find('input[name="email"]');
        emailInput.simulate('change', {
            target: {
                name: 'email',
                value: '',
            },
        });

        wrapper.find('form').simulate('submit', {
            preventDefault() { }
        });

        const [actionUi] = store.getActions();
        const expectedAction = {
            type: types.uiSetError,
            payload: msgError,
        };

        expect(actionUi).toEqual(expectedAction);
    });

    test('should display an error message if msgError contains data', () => {
        const initialState = {
            auth: {},
            ui: {
                isLoading: false,
                msgError,
            },
        };

        let store = mockStore(initialState);

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        const errorBox = wrapper.find('.auth__alert-error');

        expect(errorBox.exists()).toBeTruthy();
        expect(errorBox.text().trim()).toBe(msgError);
    });
});