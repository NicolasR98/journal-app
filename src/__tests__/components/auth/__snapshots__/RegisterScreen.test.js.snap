// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <RegisterScreen /> should display correctly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Hernando"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="example@mail.com"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="12345"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="passwordConfirm"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="12345"
    />
    <button
      className="btn btn-block btn-primary mb-5"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
